all functionalities are ready to publish

AmroSous opened the pull request:
* All functionalities are tested and ready to merge.
* All issues are solved in this version.
* We can create the first version of our application.

AmroSous committed:
Add Product class

Product is the base entity of the system.
Now we can use it in Inventory.

AmroSous committed:
Further planning and project structure

- Update UML diagram.
- Create unit testing project in the solution with xUnit.
- Add IInventory interface to decouple the implementation of service
  from the client code, this allow us to add another implementation of
  Inventory using database or file in future.
- Add three namespaces to manage project files: Models, Services, UI.

AmroSous committed:
Add greating feature

Add welcome function to program.cs
Make constructor for Application takes IInventory instance

AmroSous committed:
Merge pull request #1 from AmroSous/fet-welcome

Add greating feature

AmroSous committed:
Add product functionality added

we can now use AddProduct function in application to add a new
product into repository.

but this functionality is not tested yet

AmroSous committed:
Merge pull request #2 from AmroSous/fet-add-product

Add product functionality added

AmroSous committed:
Show all product feature completed

complete show all products and write unit test for
add product function.

AmroSous committed:
Merge pull request #3 from AmroSous/fet-get-products

Show all product feature completed

AmroSous committed:
Fix: #4#5 immutability bug

make a clone method for product classs to use it for copying.
write test for check immutability of stored and returned object
from inventory.

issues #4 #5 solved.

AmroSous committed:
Merge pull request #6 from AmroSous/fet-edit-product

Fix: #4#5 immutability bug

AmroSous committed:
Create workflow dotnet.yml

AmroSous committed:
feature edit product and find product ready

add edit feature to the program.
add search for a product using name in inventory.
write unit test to check both and immutability of objects.

AmroSous committed:
Update workflow dotnet.yml

AmroSous committed:
Merge pull request #7 from AmroSous/fet-edit-product

feature edit product and find product ready

AmroSous committed:
delete product functionality completed

now we can delete product from inventory using product name.
unit tests wrote to test it.

AmroSous committed:
Merge pull request #10 from AmroSous/fet-del-product

delete product functionality completed

AmroSous committed:
user interface ready to use with all operations.

we can now publish the first release of our application.

AmroSous committed:
Merge pull request #11 from AmroSous/user-interface

user interface ready to use with all operations.

AmroSous committed:
refactor Product class by implement ICloneable

use ICloneable interface and update its usage in other classes
remove uneccessary library impors

AmroSous committed:
fix: #8 #9 #12 ,refactor and documentation

* refactor: make Util namespace with these classes:
* IO: to handle user input/output
* OperationAbortedException: to add ability to abort operation using
--abort command
* LogType: enum to specify type of log and color of each type

AmroSous committed:
Merge pull request #13 from AmroSous/ref-enhance

fix: #8 #9 #12 ,refactor and documentation

AmroSous committed:
Add clear console option to menu

AmroSous committed:
Update workflow dotnet.yml

change branches name

AmroSous committed:
fet: console pause after showing results

AmroSous committed:
enhance unit tests

refactor tests for inventory
add unit test for find product method

AmroSous committed:
Merge pull request #14 from AmroSous/ref-enhance

refactor code and tests

AmroSous committed:
notify user about --abort command

tjaber-fts commented on SimpleInventoryManagement/Util/LogType.cs:
```diff
@@ -0,0 +1,17 @@
+     * This is an enumeration to specify type 
+     * of output, each type given a console color. 
+     */
+    public enum LogType
```
NIT: Clearly indicates the purpose of the Enum

```
public enum ConsoleColorType
{
    Information = ConsoleColor.Blue,
    Error = ConsoleColor.Red,
    Warning = ConsoleColor.DarkYellow,
    Prompt = ConsoleColor.Magenta,
    Success = ConsoleColor.DarkGreen,
    Failure = ConsoleColor.DarkGray,
    Default = ConsoleColor.White
}
```

AmroSous committed:
refactor: change LogType enum

I made LogType enum more readable as reviewer suggest
to ConsoleColorType and make enumeration values more descriptive.

AmroSous committed:
solve syntax error in enumeration values

AmroSous commented on SimpleInventoryManagement/Util/ConsoleColorType.cs:
```diff
@@ -0,0 +1,17 @@
+        Failure = ConsoleColor.DarkGray,
+        Default = ConsoleColor.White
+    }
+}
```
@tjaber-fts 
I have addressed your comment, and pushed the changes, 
please let me know if anything else need to be adjusted.

tjaber-fts commented on SimpleInventoryManagement/Program.cs:
```diff
@@ -1 +1,158 @@
+         * Show welcome message
+         * exits when user enter any key.
+         */
+        static void ShowWelcomeMessage()
```
This method currently handles two responsibilities: printing output and obtaining user input. To improve clarity and maintain single responsibility principles, let's refactor it into separate methods/functions.

ShowWelcomeMessage() -> should only print the welcome message
ShowCommands or  ShowInstructions() -> should be another method
ReadUserInput() -> Enter any key to start the application.. and take the user input


All the above methods are advice to be moved to a new class/helper i.e. `DisplayHelper.CS` or `MessageDisplayUtility.CS` or `ConsolePrompts.CS`

tjaber-fts commented on SimpleInventoryManagement/Program.cs:
```diff
@@ -1 +1,158 @@
+         * and prompt the user to number of operation and returned it. 
+         * if user enter invalid number then return -1.
+         */
+        private static int ShowMenu()
```
ShowMenu() should be void; only showing the Menu.
Read imput can be handled in a diffirent method

tjaber-fts commented on SimpleInventoryManagement/Program.cs:
```diff
@@ -1 +1,158 @@
+            ShowWelcomeMessage();
+
+            // taking inputs from user and perform operation required.
+            int op = 0;
```
NIT: Change op to operation  for better clarity and descriptiveness.

tjaber-fts commented on SimpleInventoryManagement/Services/Inventory.cs:
```diff
@@ -0,0 +1,90 @@
+     * 
+     * if the program terminated the changes made no longer exists.
+     */
+    public class Inventory : IInventory
```
NIT:  `InventoryService : IInventoryService`

tjaber-fts commented on SimpleInventoryManagement/Program.cs:
```diff
@@ -1 +1,158 @@
+                        Good bye ..
+
+                """, LogType.Info);
+            Console.ReadLine();
```
Since the user has halted the program this is unnecessary read. Let's remove it

tjaber-fts commented on SimpleInventoryManagement/Util/IO.cs:
```diff
@@ -0,0 +1,103 @@
+        public static void Log(string message, LogType type)
+        {
+            Console.ForegroundColor = (ConsoleColor)type;
+            Console.Write(message);
```
Use `WriteLine` instead of `Write` to ensure each message is on a new line, and you can remove the `"\n"` from the call."
method i.e. 
` IO.Log("Product added successfully.\n", ConsoleColorType.Success);`
Becomes
` IO.Log("Product added successfully.", ConsoleColorType.Success);`

AmroSous replied:
@tjaber-fts 
I made it like that because I need the prompt message be at the same line of user inputs.
but we can separate it into two methods: 
`Log()`  with new line. 
`LogPrompt()` without new line.
Is this a good solution?

tjaber-fts commented on SimpleInventoryManagement/Util/IO.cs:
```diff
@@ -0,0 +1,103 @@
+         * from user, if it empty or null try again.
+         * it takes message to show before take input.
+         */
+        public static string ReadNotNull(string msg)
```
Should be named:` ReadNullOrWhiteSpace(string msg)` to reflect the intent

tjaber-fts commented on SimpleInventoryManagement/Util/IO.cs:
```diff
@@ -0,0 +1,103 @@
+         * if input is --abort then
+         * throw OperationAbortedException.
+         */
+        public static string? Read()
```
Return early

```
public static string? Read()
{
    var input = Console.ReadLine()?.Trim();
    if (input == "--abort")
        throw new OperationAbortedException();
    
    return string.IsNullOrEmpty(input) ? null : input;
}
```

tjaber-fts commented on SimpleInventoryManagement/Util/IO.cs:
```diff
@@ -0,0 +1,103 @@
+     * operations, reading from user and logging (printing)
+     * messages on console.
+     */
+    public static class IO
```
Rename: `IO `-> `ConsoleIO`

tjaber-fts commented on SimpleInventoryManagement/Services/Inventory.cs:
```diff
@@ -0,0 +1,90 @@
+         * by name, and return copy of this product.
+         * if product does not exist return null
+         */
+        public Product? FindProduct(string name)
```
Can be refactored as follows:

```
public Product? FindProduct(string name)
{
    return_products.FirstOrDefault(p => p.Name.Equals(name, StringComparison.OrdinalIgnoreCase));
    
    // Not sure why you are cloning but If cloning is necessary
    return product?.Clone() as Product;

    // If cloning is not necessary
    // return product;
}
```

tjaber-fts commented on SimpleInventoryManagement/UI/Application.cs:
```diff
@@ -0,0 +1,99 @@
+        public void FindProduct()
+        {
+            string name = IO.ReadNotNull("Enter name of product: ");
+            var product = _inventory.FindProduct(name);
```
**BUG:** When product not fount (it's empty object) so the user got no message

![Screenshot_354](https://github.com/user-attachments/assets/3f0d2506-cd7e-4de2-91f8-5c37792b79f5)

You already handled it well in Edit product

```
  var product = _inventory.FindProduct(name);
            if (product == null)
                throw new InvalidOperationException("Product not found.");
```

tjaber-fts reviewed (changes requested):
Thanks AMRO; Great Job! Here's my humble feedback!

AmroSous committed:
refactor: renaming, separation of concerns

* rename classes to be more descriptive like InventoryService.
* keep all interaction logic with user in UI namespace.
* separate functions to be single responsible.
* put message printing methods in helper class in UI namespace
  "MessageDisplayHelper.cs".
* solve FindProduct empty object Bug, and show Product not found.

AmroSous committed:
refactor: don't pause the program for press when exit

AmroSous committed:
refactoring: rename methods, return fast

AmroSous committed:
last ref for requested changes in MR with main

Log method becomes two methods Log with newline
and LogPrompt without newline.

use .FirtOrDefault method in InventoryService class to
avoid creating new object when searching.

make methods return early by optimizing code in service class.

AmroSous commented:
@tjaber-fts 
I have addressed your feedback and pushed the requested changes. 

**Summary of changes:**
- Separation of concerns: 
    * Encapsulated all interaction logic into the `SimpleInventoryManagement.UI` namespace, and make entry point shorter.
    * Refactored methods to ensure single responsibility, such as those handling message display. 
- Renamed methods and classes to be more descriptive and clearer. 
- InventoryService Refatoring: 
    * Enhanced methods readability and performance, and make it return early.
    * Utilize `.FirstOrDefault() ` method for searching in `List<>` and eliminating the need for creating new `Product` object.
 
_Thank you for your valuable feedback, if there are any suggestions, please feel free to let me know._
